File: android/src/main/kotlin/com/ble_mesh/models/Message.kt
package com.ble_mesh.models

import java.util.UUID

/**
 * Type of message
 */
enum class MessageType(val value: Int) {
    PUBLIC(0),
    PRIVATE(1),
    CHANNEL(2),
    SYSTEM(3);

    companion object {
        fun fromInt(value: Int) = values().firstOrNull { it.value == value } ?: PUBLIC
    }
}

/**
 * Delivery status of a message
 */
enum class DeliveryStatus(val value: Int) {
    PENDING(0),
    SENT(1),
    DELIVERED(2),
    FAILED(3);

    companion object {
        fun fromInt(value: Int) = values().firstOrNull { it.value == value } ?: PENDING
    }
}

/**
 * Represents a message in the mesh network
 */
data class Message(
    val id: String = UUID.randomUUID().toString(),
    val senderId: String,
    val senderNickname: String,
    val content: String,
    val type: MessageType = MessageType.PUBLIC,
    val timestamp: Long = System.currentTimeMillis(),
    val channel: String? = null,
    val isEncrypted: Boolean = false,
    val status: DeliveryStatus = DeliveryStatus.PENDING,
    // Phase 2: Routing fields
    val ttl: Int = 7,              // Time-to-live (hops remaining)
    val hopCount: Int = 0,         // Number of hops taken
    val messageId: Long = MessageHeader.generateMessageId(),  // Unique message ID
    val isForwarded: Boolean = false  // Was this message forwarded?
) {
    companion object {
        /**
         * Parse message from byte array (header + payload)
         *
         * @param data Byte array containing header and payload
         * @return Message object or null if parsing fails
         */
        fun fromByteArray(data: ByteArray, senderNickname: String = "Unknown"): Message? {
            return try {
                // Parse header
                val header = MessageHeader.fromByteArray(data)

                // Extract payload
                val payloadBytes = data.copyOfRange(MessageHeader.HEADER_SIZE, data.size)
                val content = String(payloadBytes, Charsets.UTF_8)

                // Create message
                Message(
                    id = header.messageId.toString(),
                    senderId = header.senderId,
                    senderNickname = senderNickname,
                    content = content,
                    type = messageTypeFromHeaderType(header.type),
                    timestamp = System.currentTimeMillis(),
                    channel = null,
                    isEncrypted = false,
                    status = DeliveryStatus.DELIVERED,
                    ttl = header.ttl.toInt(),
                    hopCount = header.hopCount.toInt(),
                    messageId = header.messageId,
                    isForwarded = header.hopCount > 0
                )
            } catch (e: Exception) {
                android.util.Log.e("Message", "Failed to parse message from bytes", e)
                null
            }
        }

        /**
         * Convert MessageHeader type to MessageType
         */
        private fun messageTypeFromHeaderType(headerType: Byte): MessageType {
            return when (headerType) {
                MessageHeader.TYPE_PUBLIC -> MessageType.PUBLIC
                MessageHeader.TYPE_PRIVATE -> MessageType.PRIVATE
                MessageHeader.TYPE_CHANNEL -> MessageType.CHANNEL
                else -> MessageType.PUBLIC
            }
        }

        /**
         * Convert MessageType to MessageHeader type
         */
        private fun headerTypeFromMessageType(messageType: MessageType): Byte {
            return when (messageType) {
                MessageType.PUBLIC -> MessageHeader.TYPE_PUBLIC
                MessageType.PRIVATE -> MessageHeader.TYPE_PRIVATE
                MessageType.CHANNEL -> MessageHeader.TYPE_CHANNEL
                MessageType.SYSTEM -> MessageHeader.TYPE_PUBLIC
            }
        }
    }

    /**
     * Serialize message to byte array (header + payload)
     *
     * @return Byte array containing header and payload
     */
    fun toByteArray(): ByteArray {
        val payloadBytes = content.toByteArray(Charsets.UTF_8)

        val header = MessageHeader(
            type = headerTypeFromMessageType(type),
            ttl = ttl.toByte(),
            hopCount = hopCount.toByte(),
            messageId = messageId,
            senderId = senderId,
            payloadLength = payloadBytes.size.toShort()
        )

        val headerBytes = header.toByteArray()
        return headerBytes + payloadBytes
    }

    /**
     * Convert to a map for sending to Flutter
     */
    fun toMap(): Map<String, Any?> {
        return mapOf(
            "id" to id,
            "senderId" to senderId,
            "senderNickname" to senderNickname,
            "content" to content,
            "type" to type.value,
            "timestamp" to timestamp,
            "channel" to channel,
            "isEncrypted" to isEncrypted,
            "status" to status.value,
            // Phase 2: Routing fields
            "ttl" to ttl,
            "hopCount" to hopCount,
            "messageId" to messageId.toString(),
            "isForwarded" to isForwarded
        )
    }
}

File: ios/Classes/Models/Message.swift
import Foundation

/// Type of message
enum MessageType: Int {
    case publicMessage = 0
    case privateMessage = 1
    case channel = 2
    case system = 3
}

/// Delivery status of a message
enum DeliveryStatus: Int {
    case pending = 0
    case sent = 1
    case delivered = 2
    case failed = 3
}

/// Represents a message in the mesh network
struct Message {
    let id: String
    let senderId: String
    let senderNickname: String
    let content: String
    let type: MessageType
    let timestamp: Date
    let channel: String?
    let isEncrypted: Bool
    let status: DeliveryStatus
    // Phase 2: Routing fields
    let ttl: Int
    let hopCount: Int
    let messageId: Int64
    let isForwarded: Bool

    init(
        id: String = UUID().uuidString,
        senderId: String,
        senderNickname: String,
        content: String,
        type: MessageType = .publicMessage,
        timestamp: Date = Date(),
        channel: String? = nil,
        isEncrypted: Bool = false,
        status: DeliveryStatus = .pending,
        // Phase 2: Routing fields with defaults
        ttl: Int = 7,
        hopCount: Int = 0,
        messageId: Int64 = MessageHeader.generateMessageId(),
        isForwarded: Bool = false
    ) {
        self.id = id
        self.senderId = senderId
        self.senderNickname = senderNickname
        self.content = content
        self.type = type
        self.timestamp = timestamp
        self.channel = channel
        self.isEncrypted = isEncrypted
        self.status = status
        self.ttl = ttl
        self.hopCount = hopCount
        self.messageId = messageId
        self.isForwarded = isForwarded
    }

    /// Serialize message to Data (header + payload)
    func toData() -> Data {
        guard let payloadData = content.data(using: .utf8) else {
            return Data()
        }

        let header = MessageHeader(
            type: headerTypeFromMessageType(type),
            ttl: UInt8(ttl),
            hopCount: UInt8(hopCount),
            messageId: messageId,
            senderId: senderId,
            payloadLength: UInt16(payloadData.count)
        )

        let headerData = header.toData()
        return headerData + payloadData
    }

    /// Parse message from Data (header + payload)
    static func fromData(_ data: Data, senderNickname: String = "Unknown") -> Message? {
        do {
            // Parse header
            let header = try MessageHeader.fromData(data)

            // Extract payload
            let payloadData = data.subdata(in: MessageHeader.headerSize..<data.count)
            guard let content = String(data: payloadData, encoding: .utf8) else {
                return nil
            }

            // Create message
            return Message(
                id: String(header.messageId),
                senderId: header.senderId,
                senderNickname: senderNickname,
                content: content,
                type: messageTypeFromHeaderType(header.type),
                timestamp: Date(),
                channel: nil,
                isEncrypted: false,
                status: .delivered,
                ttl: Int(header.ttl),
                hopCount: Int(header.hopCount),
                messageId: header.messageId,
                isForwarded: header.hopCount > 0
            )
        } catch {
            print("Failed to parse message from data: \(error)")
            return nil
        }
    }

    /// Convert MessageHeader type to MessageType
    private static func messageTypeFromHeaderType(_ headerType: UInt8) -> MessageType {
        switch headerType {
        case MessageHeader.typePublic:
            return .publicMessage
        case MessageHeader.typePrivate:
            return .privateMessage
        case MessageHeader.typeChannel:
            return .channel
        default:
            return .publicMessage
        }
    }

    /// Convert MessageType to MessageHeader type
    private func headerTypeFromMessageType(_ messageType: MessageType) -> UInt8 {
        switch messageType {
        case .publicMessage:
            return MessageHeader.typePublic
        case .privateMessage:
            return MessageHeader.typePrivate
        case .channel:
            return MessageHeader.typeChannel
        case .system:
            return MessageHeader.typePublic
        }
    }

    /// Convert to a dictionary for sending to Flutter
    func toMap() -> [String: Any?] {
        return [
            "id": id,
            "senderId": senderId,
            "senderNickname": senderNickname,
            "content": content,
            "type": type.rawValue,
            "timestamp": Int64(timestamp.timeIntervalSince1970 * 1000),
            "channel": channel,
            "isEncrypted": isEncrypted,
            "status": status.rawValue,
            // Phase 2: Routing fields
            "ttl": ttl,
            "hopCount": hopCount,
            "messageId": String(messageId),
            "isForwarded": isForwarded
        ]
    }
}

